warehouse/admin/views/projects.py

```diff
@@ -0,0 +1,201 @@
+
+from warehouse.accounts.models import User
+from warehouse.packaging.models import Project, Release, Role, JournalEntry
+from warehouse.utils.paginate import paginate_url_factory
+
+
+@view_config(
+    route_name="admin.project.list",
+    renderer="admin/projects/list.html",
+    permission="admin",
+    uses_session=True,
+)
+def project_list(request):
+    q = request.params.get("q")
+
+    try:
+        page_num = int(request.params.get("page", 1))
+    except ValueError:
+        raise HTTPBadRequest("'page' must be an integer.") from None
+
+    projects_query = request.db.query(Project).order_by(Project.name)
+
+    if q:
+        terms = shlex.split(q)
+
+        filters = []
+        for term in terms:
+            filters.append(Project.name.ilike(term))
+
+        projects_query = projects_query.filter(or_(*filters))
+
+    projects = SQLAlchemyORMPage(
+        projects_query,
+        page=page_num,
+        items_per_page=25,
+        url_maker=paginate_url_factory(request),
+    )
+
+    return {"projects": projects, "query": q}
+
+
+@view_config(route_name="admin.project.detail",
+             renderer="admin/projects/detail.html",
+             permission="admin",
+             uses_session=True,
+             require_csrf=True,
+             require_methods=False)
+def project_detail(request):
+    try:
+        project = (
+            request.db.query(Project)
+            .filter(
+                Project.normalized_name ==
+                request.matchdict["project_name"]
+            )
+            .one()
+        )
+        maintainers = [
+            role
+            for role in (
+                request.db.query(Role)
+                .join(User)
+                .filter(Role.project == project)
+                .distinct(User.username)
+                .all()
+            )
+        ]
+        maintainers = sorted(
+            maintainers,
+            key=lambda x: (x.role_name, x.user.username),
+        )
+        journal = [
+            entry
+            for entry in (
+                request.db.query(JournalEntry)
+                .filter(JournalEntry.name == project.normalized_name)
+                .order_by(JournalEntry.submitted_date.desc())
+                .limit(50)
+            )
+        ]
+    except NoResultFound:
+        raise HTTPNotFound
+
+    return {"project": project, "maintainers": maintainers, "journal": journal}
+
+
+@view_config(
+    route_name="admin.project.releases",
+    renderer="admin/projects/releases_list.html",
+    permission="admin",
+    uses_session=True,
+)
+def releases_list(request):
+    q = request.params.get("q")
+    project_name = request.matchdict["project_name"]
+
+    try:
+        page_num = int(request.params.get("page", 1))
+    except ValueError:
+        raise HTTPBadRequest("'page' must be an integer.") from None
+
+    project = (request.db.query(Project)
+               .filter(or_(Project.name == project_name,
```

Comment:
Project detail only works with the normalized name, this view works with either. Probably we should be consistent one way or the other. You can actually use the ``ProjectFactory`` to do that though. Take a look at th enormal project detail pages to see how that works. It also includes redirecting to the canonical URL which we probably want to do here too.

Reply:
Alternatively we might want to consider adding a UUID pk to these URLs and using that in the admin like was done for Users. That way our admin URLs are stable even if something gets renamed etc. Downside to that of course is that our Admin URLs are not as "pretty" and don't encode anything human readable in them.

Reply:
@dstufft good tip, but the ProjectFactory approach appears to be interfering with the permissions. any pointers? i'm not really all that familiar with pyramid and it's terminologies.

Reply:
Oh, I think you need to add the admin permission to the Admin group for the Project. In the Model for Projects there is a ``__acl__`` method which returns a list of ACLs, you need to add one like: ``(Allow, "group:admins", "admin")``, which will say to Allow anyone with "group:admins" principal (e.g. if you're a site admin you have that principal assigned to you) actions on this object that require the ``"admin"`` permission.

Reply:
In Pyramid, views have contexts, so what you're doing with that ProjectFactory is setting the context for that view, and the permission system uses the ACLs of the context to decide if the current user has permissions to do this view or not.

It works by default, because Pyramid has a default No-op context, which exists at https://github.com/pypa/warehouse/blob/c6055f056faa6170706bc54c691be231b26c17eb/warehouse/config.py#L54-L64 that I setup to already have that ACL added.